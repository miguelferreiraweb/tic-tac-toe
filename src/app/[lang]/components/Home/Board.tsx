import React from 'react';
import { v4 as uuidv4 } from 'uuid';

import styles from '@/styles/pages/Home/components/Board.module.scss';
import { BOARD_CELLS } from '@/utils/constants/board';
import { BoardSymbolEnum, BoardSymbolType } from '@/utils/entities/board';

interface BoardProps {
  board: BoardSymbolType[];
  handleCellClick: (_arg: number) => void;
}

export default function Board({ board, handleCellClick }: BoardProps) {
  return (
    <div className={styles.boardContainer}>
      {[...Array(BOARD_CELLS)].map((_item, i: number) => (
        <button
          key={uuidv4()}
          className={`${styles.cell} ${styles[`cell--${i}`]}`}
          onClick={() => handleCellClick(i)}
        >
          <span className={board[i] === BoardSymbolEnum.PlayerX ? styles.playerX : styles.playerO}>
            {board[i]}
          </span>
        </button>
      ))}
    </div>
  );
}
