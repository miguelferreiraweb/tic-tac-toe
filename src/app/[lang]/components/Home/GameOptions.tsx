import clsx from 'clsx';
import { useTranslations } from 'next-intl';
import React, { useEffect, useState } from 'react';

import { useGame } from '@/hooks/useGame';
import {
  getFinishRoundAction,
  getUpdateBoardAction,
  getUpdateCurrentPlayerAction,
} from '@/store/Game/gameActions';
import styles from '@/styles/pages/Home/components/GameOptions.module.scss';
import {
  BOARD_INITIAL_STATE,
  INITIAL_RESTART_COUNTER,
  INTERVAL_TIMER_MS,
  LOCALES,
} from '@/utils/constants/board';
import { BoardSymbolEnum } from '@/utils/entities/board';

export default function GameOptions() {
  const {
    state: { isRoundFinished, board, currentPlayer },
    dispatch,
  } = useGame();
  const [restartCounter, setRestartCounter] = useState<number>(INITIAL_RESTART_COUNTER);
  const t = useTranslations(LOCALES.HOME);

  useEffect(() => {
    let intervalId: NodeJS.Timer;

    if (isRoundFinished) {
      // runs the interval every second and decrements the restart counter.
      // when counter reaches 0, it will reset/restart the game.
      intervalId = setInterval(() => {
        setRestartCounter(restartCounter - 1);

        if (restartCounter <= 0) {
          resetGame();
          setRestartCounter(INITIAL_RESTART_COUNTER);
        }
      }, INTERVAL_TIMER_MS);
    }

    return () => clearInterval(intervalId);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [isRoundFinished, restartCounter]);

  const resetGame = (): void => {
    dispatch(getUpdateBoardAction(BOARD_INITIAL_STATE));
    dispatch(
      getUpdateCurrentPlayerAction(
        currentPlayer === BoardSymbolEnum.PlayerX
          ? BoardSymbolEnum.PlayerO
          : BoardSymbolEnum.PlayerX
      )
    );
    dispatch(getFinishRoundAction(false));
  };

  const handleRestartGameClick = (): void => {
    if (!isBoardEmpty() && !isRoundFinished) {
      resetGame();
    }
  };

  const isBoardEmpty = (): boolean =>
    !board.includes(BoardSymbolEnum.PlayerX) && !board.includes(BoardSymbolEnum.PlayerO);

  return (
    <div className={styles.gameOptionsContainer}>
      <button
        onClick={handleRestartGameClick}
        className={clsx(styles.restartBtn, {
          [styles['restartBtn--disabled']]: isBoardEmpty(),
          [styles['restartBtn--reset']]: isRoundFinished,
        })}
        disabled={isBoardEmpty()}
      >
        {isRoundFinished ? (
          <>
            {t('restarting')} {restartCounter}
          </>
        ) : (
          t('restart')
        )}
      </button>
    </div>
  );
}
