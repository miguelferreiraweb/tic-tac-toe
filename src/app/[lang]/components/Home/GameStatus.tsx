import { useTranslations } from 'next-intl';
import React from 'react';

import { useGame } from '@/hooks/useGame';
import styles from '@/styles/pages/Home/components/GameStatus.module.scss';
import { LOCALES } from '@/utils/constants/board';
import { RoundStatusEnum } from '@/utils/entities/board';

export default function GameStatus() {
  const {
    state: { isRoundFinished, currentPlayer, roundStatus },
  } = useGame();
  const t = useTranslations(LOCALES.HOME);

  const renderRoundStatusMsg = (): JSX.Element =>
    roundStatus === RoundStatusEnum.Draw ? (
      <span>{t('its-a-draw')}</span>
    ) : (
      <span className={styles.winnerText}>
        {t('player')}&nbsp;{currentPlayer} {t('won')}!
      </span>
    );

  return (
    <div className={styles.statusContainer}>
      {isRoundFinished ? (
        renderRoundStatusMsg()
      ) : (
        <>
          {t('its-turn-of')} {currentPlayer}
        </>
      )}
    </div>
  );
}
