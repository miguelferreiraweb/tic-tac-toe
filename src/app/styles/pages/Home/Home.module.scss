.homeContainer {
  display: grid;
  place-content: center;
  text-align: center;
  height: 100%;
}
