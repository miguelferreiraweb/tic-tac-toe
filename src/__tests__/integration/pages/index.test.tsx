import { render } from '@testing-library/react';
import { describe, expect, it } from 'vitest';

import Home from '@/pages/index';

describe('Home', () => {
  it('should render homepage unchanged', () => {
    const { container } = render(<Home />);
    expect(container).toMatchSnapshot();
  });
});
