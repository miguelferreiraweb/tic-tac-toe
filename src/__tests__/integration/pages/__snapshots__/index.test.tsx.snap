// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Home > should render homepage unchanged 1`] = `
<div>
  <div
    class="_homeContainer_b18682"
  >
    <main
      class="_main_b18682"
    >
      <div
        class="_start_b18682"
      >
        its-turn-of
         
        !
      </div>
      <div
        class="_boardContainer_b18682"
      >
        <button
          class="_cell_b18682 _cell--0_b18682"
        >
          <span
            class="_playerO_b18682"
          />
        </button>
        <button
          class="_cell_b18682 _cell--1_b18682"
        >
          <span
            class="_playerO_b18682"
          />
        </button>
        <button
          class="_cell_b18682 _cell--2_b18682"
        >
          <span
            class="_playerO_b18682"
          />
        </button>
        <button
          class="_cell_b18682 _cell--3_b18682"
        >
          <span
            class="_playerO_b18682"
          />
        </button>
        <button
          class="_cell_b18682 _cell--4_b18682"
        >
          <span
            class="_playerO_b18682"
          />
        </button>
        <button
          class="_cell_b18682 _cell--5_b18682"
        >
          <span
            class="_playerO_b18682"
          />
        </button>
        <button
          class="_cell_b18682 _cell--6_b18682"
        >
          <span
            class="_playerO_b18682"
          />
        </button>
        <button
          class="_cell_b18682 _cell--7_b18682"
        >
          <span
            class="_playerO_b18682"
          />
        </button>
        <button
          class="_cell_b18682 _cell--8_b18682"
        >
          <span
            class="_playerO_b18682"
          />
        </button>
      </div>
      <div
        class="_restart_b18682"
      >
        <button
          class="_restartBtn_b18682 _restartBtn--disabled_b18682"
          disabled=""
        >
          restart
        </button>
      </div>
    </main>
  </div>
</div>
`;
